util-app/src/main/scala/com/twitter/app/ClassPath.scala

```diff
@@ -51,84 +52,81 @@ private[app] sealed abstract class ClassPath[CpInfo <: ClassPath.Info] {
   protected def ignoredPackages: Set[String]
 
   def browse(loader: ClassLoader): Seq[CpInfo] = {
-    val buf = mutable.Buffer[CpInfo]()
+    val buf = Vector.newBuilder[CpInfo]
     val seenUris = mutable.HashSet[URI]()
 
     for ((uri, loader) <- getEntries(loader)) {
       browseUri0(uri, loader, buf, seenUris)
       seenUris += uri
     }
-    buf
+    buf.result
   }
 
   // package protected for testing
   private[app] def getEntries(loader: ClassLoader): Seq[(URI, ClassLoader)] = {
-    val ents = mutable.Buffer[(URI, ClassLoader)]()
-    val parent = loader.getParent
-    if (parent != null)
-      ents ++= getEntries(parent)
-
-    loader match {
-      case urlLoader: URLClassLoader =>
-        Option(urlLoader.getURLs) match {
-          case Some(urls) =>
-            urls.foreach { url =>
-              if (url != null)
-                ents += (url.toURI -> loader)
-            }
-          case _ =>
-        }
-      case _ =>
-    }
+    val parent = Option(loader.getParent)
+ 
+    val ownURIs: Vector[(URI, ClassLoader)] = for {
+      urlLoader <- Vector(loader).collect {case u: URLClassLoader => u}
+      urls <- Option(urlLoader.getURLs()).toVector
+      url <- urls if url != null
+    } yield (url.toURI -> loader)
+
+    val p = parent.toSeq.flatMap(getEntries)
+
+    ownURIs ++ p
 
-    ents
   }
 
   // package protected for testing
-  private[app] def browseUri(uri: URI, loader: ClassLoader, buf: mutable.Buffer[CpInfo]): Unit =
+  private[app] def browseUri(uri: URI, loader: ClassLoader, buf: Builder[CpInfo, Seq[CpInfo]]): Unit =
     browseUri0(uri, loader, buf, mutable.Set[URI]())
 
   private[this] def browseUri0(
     uri: URI,
     loader: ClassLoader,
-    buf: mutable.Buffer[CpInfo],
+    buf: Builder[CpInfo, Seq[CpInfo]],
     history: mutable.Set[URI]
   ): Unit = {
-    if (uri.getScheme != "file")
-      return
-
-    val f = new File(uri)
-    if (!(f.exists() && f.canRead))
-      return
-
-    if (f.isDirectory)
-      browseDir(f, loader, "", buf)
-    else
-      browseJar(f, loader, buf, history)
+    if (!history.contains(uri)) {
```

Comment:
do URI checking local to this function -- forking in test resulted in duplicate CpInfo's in the result